modules/data.atmosphere/R/met_temporal_downscale.Gaussian_ensemble.R

```diff
@@ -0,0 +1,398 @@
+  # Reading in the source data
+  source <- list()
+  tem <- ncdf4::nc_open(input_met)
+  for (j in seq_along(var$CF.name)) {
+    if (exists(as.character(var$CF.name[j]), tem$var) == FALSE) {
+      source[[j]] <- NA
+    } else {
+      source[[j]] <- ncdf4::ncvar_get(tem, as.character(var$CF.name[j]))
+    }
+  }
+  ncdf4::nc_close(tem)
+  
+  source <- data.frame(source)
+  colnames(source) <- var$CF.name
+  
+  # Default downscale will be to the resolution of the training dataset
+  reso <- 24/(nrow(train)/365)
+  reso_len <- nrow(train)
+  
+  # If a source doesn't have leap days, we need to remove them to ensure equal
+  # lengths
+  if (lubridate::leap_year(year) == TRUE) 
+  {
+    if (length(source$air_temperature)%%366 > 0) {
+      if (length(train$air_temperature)%%366 == 0) {
+        train <- train[1:365 * (nrow(train)/366)]
+      }
+      eph_year <- 2006
+    }
+  }  #chose a non-leap year to use for daylength calculations if we don't have the 
+  if (lubridate::leap_year(eph_year) == TRUE) {
+    sp <- 366
+  }
+  if (lubridate::leap_year(eph_year) == FALSE) {
+    sp <- 365
+  }
+  
+  # Now we start a for loop for the ensemble members and begin downscaling. A
+  # random normal distribution is used to downscale as so;
+  # (mean <- value of source data) (sd <- +/- window_days of train data at the
+  # same time intervals) 
+  for (e in seq_along(1:ensemble_members)) {
+    
+    div <- nrow(train)/nrow(source)  #tells us how many values need to be generated (via downscaling) from each source value
+    sd_step <- nrow(train)/sp  #allows us to step through each window at specific times
+    
+    # Temperature For temperature, we create a random normal distribution with the
+    # mean of the source value and the standard deviation of the training value
+    # within the window, and step through that window sequentially so that we are
+    # only using the standard devaition of values recorded at the same time as the
+    # value we are generating.
+    temp <- c("air_temperature", "air_temperature_max", "air_temperature_min")
+    df <- data.frame()
+    for (u in temp) {
+      woo <- vector()
+      a <- as.numeric(train[[u]])
+      sour <- as.numeric(source[[u]])
+      if (all(is.na(sour)) == TRUE) {
+        woo <- rep(NA, reso_len)
+      }
+      if (all(is.na(sour)) == FALSE) {
+        for (x in seq_along(sour)) {
+          four <- vector()
+          for (n in seq_along(1:div)) {
+            lowday <- x * div - wd * sd_step + n
+            highday <- x * div + wd * sd_step + n
+            if (lowday < 0) {
+              lowday <- x + n
+              highday <- x + n + wd * sd_step + n
+            }
+            if (highday > reso_len) {
+              highday <- reso_len
+              lowday <- x * div - wd * sd_step + n
+            }
+            four[n] <- rnorm(1, mean = sour[x], sd = sd(a[seq(from = lowday, 
+                                                              to = highday, by = sd_step)]))
```

Comment:
Trying to figure out what you're doing here. So I'm imagining, for example, downscaling a specific half hour within a 6 hour chunk of time from the source. First you're setting the mean temperature to be the temperature of the source, which will result in the downscaled met behaving like a step function since all 12 half hours in a 6hr block will have the same mean. That doesn't seem right to me -- it seems like we should be doing some sort of interpolation or some sort of correction that accounts for the anomalies in the training data (e.g. the 3rd half hour is always x% warmer than average across the training window).  The other thing that doesn't seem right is that the downscaling is only looking within a single source time step, so even if you fix the first problem you would still end up with discontinuities between time steps.  Next, for the standard deviation it looks like you're subsetting the training data that's within a window +/- some number of days around the current time and calculating it's standard deviation. If I understand correct, you are at least making sure you select the correct time of day. However, what you're doing fundamentally confuses uncertainty and variability. The SD you are calculating is a measure of how weather changes from day-to-day, not the uncertainty in the downscaling algorithm, which needs to be calculated as some sort of residual between the downscaled met and the observed met. Furthermore, the draws for any two time steps are independent, which is going to result in a driver that is very very noisy -- not only is that not how the world works, but it's going to blow up the land surface models. Finally, there seems to be a major time indexing problem. If, for example, my training data is 2 years long and my source data is 100 years long, when x is in years 3:100  then highday > reso_len, which implies every single day in every single year thereafter is downscaled  based on Dec 31 of year 2, and lowday will end up > reso_len, which will throw an error.